doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,102 +1,126 @@
 
-These tactics use the equality :g:`eq:forall A:Type, A->A->Prop` defined in
-file ``Logic.v`` (see :ref:`coq-library-logic`). The notation for :g:`eq T t u` is
-simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
+   .. coqdoc::
 
-.. tacn:: rewrite @term
-   :name: rewrite
+      Inductive eq {A : Type} (x : A) : A -> Prop := eq_refl : eq x x.
 
-   This tactic applies to any goal. The type of :token:`term` must have the form
+  The notation `x = y` represents the term `eq x y`.  The notation `x = y :> A`
+  gives the type of x and y explicitly.
 
-   ``forall (x``:sub:`1` ``:A``:sub:`1` ``) ... (x``:sub:`n` ``:A``:sub:`n` ``), eq term``:sub:`1` ``term``:sub:`2` ``.``
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support.
 
-   where :g:`eq` is the Leibniz equality or a registered setoid equality.
+- :gdef:`Definitional equality <definitional equality>` is equality based on the
+  :ref:`conversion rules <Conversion-rules>`, which Coq can determine automatically.
+  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
+  two terms are definitionally equal.
 
-   Then :n:`rewrite @term` finds the first subterm matching `term`\ :sub:`1` in the goal,
-   resulting in instances `term`:sub:`1`' and `term`:sub:`2`' and then
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+.. _rewritingexpressions:
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+Rewriting with Leibniz and setoid equality
+------------------------------------------
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-   .. tacv:: rewrite -> @term
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-      Is equivalent to :n:`rewrite @term`
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-   .. tacv:: rewrite <- @term
+   Rewrites terms based on equalities.  :n:`@one_term` is a term whose type has the form:
```

Comment:
"one_term is a term" seems a bit obvious. What about "The type of one_term should have the form"?